stan/math/prim/fun/matrix_exp.hpp

```diff
@@ -12,19 +13,25 @@ namespace math {
  * Return the matrix exponential of the input
  * matrix.
  *
- * @tparam T type of elements in the matrix
- * @param[in] A Matrix to exponentiate.
+ * @tparam T type of the matrix
+ * @param[in] A_in Matrix to exponentiate.
  * @return Matrix exponential, dynamically-sized.
  * @throw <code>std::invalid_argument</code> if the input matrix
  * is not square.
  */
-template <typename T>
-inline Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> matrix_exp(
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic>& A) {
+template <typename T, typename = require_eigen_t<T>>
+inline plain_type_t<T> matrix_exp(const T& A_in) {
+  using std::exp;
+  const auto& A = A_in.eval();
```

Comment:
Yes. 
```
const auto& a = b.eval();
```
will evaluate any expression `b`.

```
const Eigen::Ref<const typename T::PlainObject>& a = b;
```
will evaluate if `b`'s elements can not be directley accessed. In other words if `b` is something like `c+d` or `3*c` it will evaluate it. If `b` is something like `c.block(...).transpose()` it will not be evaluated.

```
const auto& a = b;
```
This will never evaluate anything. `a` will just be a reference to `b`, regardless of what `b` is.

Reply:
Actually I should use `Ref` here.

Reply:
Ty for the 'splainer!

Side note wish they had some way to do Ref but on the rhs so auto& and auto&& could be used in these cases